### PR TITLE
Addition of Siamese loss or contrastive loss function #34879

### LINKED ISSUE #34879: Addition of Siamese loss or contrastive loss function
## 🚀 Feature

I would like to contribute to the pytorch by adding the [ contrastive loss](http://yann.lecun.com/exdb/publis/pdf/hadsell-chopra-lecun-06.pdf) function which is usually implemented  for siamese networks

## Motivation

I have been working on siamese networks and have implemented constrastive loss to measure similarity between bottleneck embeddings of the autoencoder.

## Pitch

The output of the loss function is similar to that of  [this](https://github.com/pytorch/pytorch/blob/d9b97a4ffd3d65eeaccb5f8729926976fa868981/torch/nn/functional.py#L3544) but for siamese network instead of triplet network.

It returns a float value signifying the distance between the bottleneck embedding.

## Additional context

This loss implemented for all siamese network implementations (such as facial similarity, Omniglot examples)

### PR DESCRIPTION
**Pull request by Hemantr05**
Fixes #34879
Have added the siamese loss as mentioned in issue #34879, for python.

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -3345,6 +3345,13 @@
   dispatch:
     DefaultBackend: celu_
 
+- func: siamese_loss(Tensor input1, Tensor input2, Tensor label, float margin=1.0, int reduction=Mean) -> Tensor
+  use_c10_dispatcher: fill
+  python_module: nn
+  dispatch:
+    CPU: siamese_loss
```
Comment (jbschlosser) [inline aten/src/ATen/native/native_functions.yaml:3352]: I think the implementations for these functions are missing from this PR?

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -3345,6 +3345,13 @@
   dispatch:
     DefaultBackend: celu_
 
+- func: siamese_loss(Tensor input1, Tensor input2, Tensor label, float margin=1.0, int reduction=Mean) -> Tensor
+  use_c10_dispatcher: fill
```
Comment (jbschlosser) [inline aten/src/ATen/native/native_functions.yaml:3349]: This line is no longer needed, as the full c10 dispatcher is used by default.

File: torch/nn/modules/loss.py [hunk 3]
```diff
@@ -1538,6 +1538,79 @@ def forward(self, anchor: Tensor, positive: Tensor, negative: Tensor) -> Tensor:
...
+    """
+    __constants__ = ['margin', 'reduction']
+    margin: float
+
+    def __init__(self, margin: float=1.0, size_average = None, reduce = None, reduction: str = 'mean'):
+        super(SiameseLoss, self).__init__(size_average, reduce, reduction)
+        self.margin = margin
+
+    def forward(self, anchor: Tensor, positive: Tensor, label: Tensor) -> Tensor:
+        return F.siames_loss(anchor, positive, margin = self.margin, reduction = self.reduction)
```
Comment (jbschlosser) [inline torch/nn/modules/loss.py:1612]:
```suggestion
        return F.siamese_loss(anchor, positive, margin = self.margin, reduction = self.reduction)
```

**Review by jbschlosser — COMMENTED** [review]
Thanks for the contribution! Is there more coming for this PR?

In particular, the C++ implementation and unit tests are missing.

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (Hemantr05) [reply aten/src/ATen/native/native_functions.yaml:3349]: Oh alright.
Wouldn't line 3348 be necessary though?

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (Hemantr05) [reply aten/src/ATen/native/native_functions.yaml:3352]: I have added the functions in torch/nn/functional.py

**Comment by Hemantr05** [comment]
@jbschlosser Yes, I will be implementing the c++ and unit tests as well

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (jbschlosser) [reply aten/src/ATen/native/native_functions.yaml:3349]: Yep, line 3348 looks good!

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (jbschlosser) [reply aten/src/ATen/native/native_functions.yaml:3352]: Quick clarification (I know there's a lot and it can be confusing) - the function in `torch/nn/functional.py` is the python implementation, and, in most cases, that just calls into a C++ implementation (alongside some size checks and `handle_torch_function()` logic like you've added). It looks like your python implementation currently calls:

```python
return torch.siamese_loss(input1, input2, label, margin)
```

The entry in `native_functions.yaml` you've added correctly declares a `siamese_loss()` function that python can call into, but there's a couple issues:
1. Since you specified `python_module: nn` above (which I agree with), the function will actually be available in python as `torch._C.nn.siamese_loss()`.
2. This function still needs C++ definitions underneath `aten/src/ATen/native`. The `CPU` and `CUDA` dispatch entries indicate which C++ functions define the implementation for each device.

So to be precise, for this `CPU: siamese_loss` line, I'd expect a C++ function named `siamese_loss()` that matches the interface you've specified above and implements the core logic, and it should likely be in `aten/src/ATen/native/Loss.cpp`.

Hopefully that all makes sense! Glad to help out more if needed :)

Comment (Hemantr05) [reply aten/src/ATen/native/native_functions.yaml:3352]: Makes sense, will implement the c++ function that matches the interface. Will definitely reach out if there's anything
